os/lib/sha-256.c

```diff
@@ -0,0 +1,461 @@
+      | ((uint32_t)p[2] << 8)
+      | p[3];
+}
+
+static void
+be32enc(uint8_t *p, uint32_t u)
+{
+  p[0] = (u >> 24) & 0xff;
+  p[1] = (u >> 16) & 0xff;
+  p[2] = (u >> 8) & 0xff;
+  p[3] = u & 0xff;
+}
+
+static void
+be64enc(uint8_t *p, uint64_t u)
+{
+  be32enc(p, (uint32_t)(u >> 32));
+  be32enc(p + 4, (uint32_t)(u & 0xffffffffU));
+}
+
+/*
+ * Encode a length len/4 vector of (uint32_t) into a length len vector of
+ * (unsigned char) in big-endian form.  Assumes len is a multiple of 4.
+ */
+static void
+be32enc_vect(uint8_t *dst, const uint32_t *src, size_t len)
+{
+  size_t i;
+
+  for(i = 0; i < len / 4; i++) {
+    be32enc(dst + i * 4, src[i]);
+  }
+}
+
+/*
+ * Decode a big-endian length len vector of (unsigned char) into a length
+ * len/4 vector of (uint32_t).  Assumes len is a multiple of 4.
+ */
+static void
+be32dec_vect(uint32_t *dst, const uint8_t *src, size_t len)
+{
+  size_t i;
+
+  for(i = 0; i < len / 4; i++) {
+    dst[i] = be32dec(src + i * 4);
+  }
+}
+#endif /* UIP_BYTE_ORDER != UIP_LITTLE_ENDIAN */
+
+static const uint32_t K[64] = {
+  0x428a2f98, 0x71374491, 0xb5c0fbcf, 0xe9b5dba5,
+  0x3956c25b, 0x59f111f1, 0x923f82a4, 0xab1c5ed5,
+  0xd807aa98, 0x12835b01, 0x243185be, 0x550c7dc3,
+  0x72be5d74, 0x80deb1fe, 0x9bdc06a7, 0xc19bf174,
+  0xe49b69c1, 0xefbe4786, 0x0fc19dc6, 0x240ca1cc,
+  0x2de92c6f, 0x4a7484aa, 0x5cb0a9dc, 0x76f988da,
+  0x983e5152, 0xa831c66d, 0xb00327c8, 0xbf597fc7,
+  0xc6e00bf3, 0xd5a79147, 0x06ca6351, 0x14292967,
+  0x27b70a85, 0x2e1b2138, 0x4d2c6dfc, 0x53380d13,
+  0x650a7354, 0x766a0abb, 0x81c2c92e, 0x92722c85,
+  0xa2bfe8a1, 0xa81a664b, 0xc24b8b70, 0xc76c51a3,
+  0xd192e819, 0xd6990624, 0xf40e3585, 0x106aa070,
+  0x19a4c116, 0x1e376c08, 0x2748774c, 0x34b0bcb5,
+  0x391c0cb3, 0x4ed8aa4a, 0x5b9cca4f, 0x682e6ff3,
+  0x748f82ee, 0x78a5636f, 0x84c87814, 0x8cc70208,
+  0x90befffa, 0xa4506ceb, 0xbef9a3f7, 0xc67178f2,
+};
+
+/* Elementary functions used by SHA-256 */
+#define Ch(x, y, z)  ((x & (y ^ z)) ^ z)
+#define Maj(x, y, z) ((x & (y | z)) | (y & z))
+#define SHR(x, n)    (x >> n)
+#define ROTR(x, n)   ((x >> n) | (x << (32 - n)))
+#define S0(x)        (ROTR(x, 2) ^ ROTR(x, 13) ^ ROTR(x, 22))
+#define S1(x)        (ROTR(x, 6) ^ ROTR(x, 11) ^ ROTR(x, 25))
+#define s0(x)        (ROTR(x, 7) ^ ROTR(x, 18) ^ SHR(x, 3))
+#define s1(x)        (ROTR(x, 17) ^ ROTR(x, 19) ^ SHR(x, 10))
+
+/* SHA256 round function */
+#define RND(a, b, c, d, e, f, g, h, k) \
+  h += S1(e) + Ch(e, f, g) + k; \
+  d += h; \
+  h += S0(a) + Maj(a, b, c);
+
+/* Adjusted round function for rotating state */
+#define RNDr(S, W, i, ii) \
+  RND(S[(64 - i) % 8], S[(65 - i) % 8], \
+      S[(66 - i) % 8], S[(67 - i) % 8], \
+      S[(68 - i) % 8], S[(69 - i) % 8], \
+      S[(70 - i) % 8], S[(71 - i) % 8], \
+      W[i + ii] + K[i + ii])
+
+/* Message schedule computation */
+#define MSCH(W, ii, i) \
+  W[i + ii + 16] = \
+      s1(W[i + ii + 14]) + W[i + ii + 9] + s0(W[i + ii + 1]) + W[i + ii]
+
+static sha_256_checkpoint_t ctx;
```

Comment:
There are many different contexts in Contiki-NG, so a slightly more unique name than `ctx` would be beneficial.

Does prepending `sha_256_` (or `sha_`, or some other suitable prefix/suffix) make the lines too long and make the file unreadable?

Reply:
Renamed this variable to `checkpoint`.